sources/brush/src/main.rs

```diff
@@ -0,0 +1,429 @@
+            symlink {ALLOWED_PROGRAMS_DIR}/s3-mountpoint /opt/csi/bin/s3-mountpoint
+        "#
+        );
+        populate_directory(&tmpdir, &test_layout);
+
+        macro_rules! tmp {
+            ($path:expr) => {
+                path_inside(&tmpdir, $path)
+            };
+        }
+
+        let make_paths = |p: Vec<&str>| {
+            let mut s = String::new();
+            for x in p.iter() {
+                s.push_str(&format!("{}:", tmp!(x).display()));
+            }
+            s
+        };
+
+        let allowed_programs = tmp!(ALLOWED_PROGRAMS_DIR);
+        let normal_paths = make_paths(vec!["/usr/sbin", "/usr/bin"]);
+
+        // These all resolve to `coreutils` via symlinks.
+        for allowed in ["true", "false", "chroot"] {
+            assert_eq!(
+                verify_program(&allowed_programs, &normal_paths, allowed).unwrap(),
+                (tmp!("/usr/bin/coreutils"), filename(allowed).unwrap())
+            );
+        }
+
+        // These also resolve to `coreutils`, but no symlink exists to allow them.
+        for not_allowed in ["rm", "date"] {
+            assert!(verify_program(&allowed_programs, &normal_paths, not_allowed).is_err());
+        }
+
+        let other_paths = make_paths(vec!["/opt/cni/bin", "/opt/csi/bin"]);
+
+        // These programs are not in the normal paths, but should be located in other paths.
+        for other in ["cnitool", "s3-mountpoint"] {
+            assert!(verify_program(&allowed_programs, &normal_paths, other).is_err());
+            assert!(verify_program(&allowed_programs, &other_paths, other).is_ok());
+        }
+
+        let all_paths = make_paths(vec!["/usr/bin", "/opt/cni/bin"]);
+
+        // This succeeds because the absolute path for `cnitool` avoids the $PATH lookup.
+        assert!(
+            verify_program(&allowed_programs, &all_paths, tmp!("/opt/cni/bin/cnitool")).is_ok()
+        );
+
+        // This is an error because it finds `cnitool` in /usr/bin, but the allowed symlink points
+        // to /opt/cni/bin.
+        assert!(verify_program(&allowed_programs, &all_paths, "cnitool").is_err());
+    }
+
+    // Helper to build input to parse_args()
+    fn args_from(s: &[&str]) -> Vec<String> {
+        s.iter().map(|s| s.to_string()).collect()
+    }
+
+    // Helper to build return value from parse_args()
+    fn cmd_from(s: &[&str]) -> (String, Vec<String>) {
+        let mut s = s.iter();
+        let program = s.next().unwrap().to_string();
+        let args = s.map(|s| s.to_string()).collect();
+        (program, args)
+    }
+
+    #[test]
+    fn test_parse_args() {
+        let good = args_from(&["sh", "-c", "qwerty --uiop asdf"]);
+        let expected = cmd_from(&["qwerty", "--uiop", "asdf"]);
+        assert_eq!(parse_args(good).unwrap(), expected);
+
+        let bad = args_from(&["sh", "zxcvb"]);
+        assert!(parse_args(bad).is_err());
+
+        let bad = args_from(&["sh", "-c"]);
+        assert!(parse_args(bad).is_err());
```

Comment:
nit: I'd like to catch the other case I could see a user doing where they pass all of the arguments as separate args and we want to explicitly signal this isn't a bug but one should pass the entire list as a single string for the program execution.
```suggestion
        assert!(parse_args(bad).is_err());
        
        let bad = args_from(&["sh", "-c", "querty", "--uiop", "asdf"]);
        assert!(parse_args(bad).is_err());
```

Reply:
Added this test case.